### PR TITLE
Add OutputPath metadata to PublishItemsOutputGroupOutputs

### PR DESCRIPTION
**Pull request by wli3**
To resolve wrapproj logic duplication, create a new metadata
“OutputPath” on  PublishItemsOutputGroupOutputs at the end of Publish
target to represent the content of PublishDir. By the time the itemgroup
is created, the files in this itemgroup should exist on disk. So that
wapproj target do not need to prob the content of PublishDir.

Wapproj targets should depend on Publish target with additional props
being set. And it should set PublishDir to desire end output directory
to avoid an extra copying.

image

### CODE REVIEW

**Review by MSLukeWest — COMMENTED** [review]
Can/should your new PublishItemsOutputGroupOutputsTests.cs be combined with the existing PublishItemsOutputGroupTests.cs?  It seems like they do some similar verification

**Comment by wli3** [comment]
@MSLukeWest I am replicating _FilesCopiedToPublishDir_ tests for _PublishItemsOutputGroupOutputsTests_, since _PublishItemsOutputGroupOutputsTests_ would replace _FilesCopiedToPublishDir_

File: src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                DependsOnTargets = "Publish",
+                MetadataNames = { "OutputPath" },
+            };
+
+            command.Execute().Should().Pass();
+            var items = from item in command.GetValuesWithMetadata()
+                        select new
+                        {
+                            Identity = item.value,
+                            OutputPath = Path.GetFileName(Path.GetFullPath(Path.Combine(testAsset.Path, testProject.Name, item.metadata["OutputPath"])))
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs:82]: Why the call to `Path.GetFullPath` here if it's just going to strip off everything with the call to `Path.GetFileName`?

```suggestion
                            OutputPath = Path.GetFileName(item.metadata["OutputPath"])
```

File: src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            // Framework assemblies should be there if we specified and rid and this isn't in the single file case
+            if (specifyRid && !singleFile)
+            {
+                FrameworkAssemblies.ForEach(fa => items.Should().ContainSingle(i => i.OutputPath.Equals(fa, StringComparison.OrdinalIgnoreCase)));
+                items.Should().Contain(i => i.OutputPath.Equals($"{testProject.Name}.deps.json", StringComparison.OrdinalIgnoreCase));
+            }
+            else
+            {
+                FrameworkAssemblies.ForEach(fa => items.Should().NotContain(i => i.OutputPath.Equals(fa, StringComparison.OrdinalIgnoreCase)));
+
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs:107]: Nit: delete whitespace

**Review by dsplaisted — APPROVED** [review]
Looks good.  It would be nice if we could start to document more about how these publish items, metadata, and targets are used.  For example, why do we have both `FilesCopiedToPublishDir` as well as `PublishItemsOutputGroupOutputs`.  What are the different consumers?

File: src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs (hunk 1 above)
Comment (wli3) [reply src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs:82]: Directly call FullPath will put current PWD in front, which is wrong. It is more correct even for an intermediate value to combine the right path first.

Comment (dsplaisted) [reply src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs:82]: But the call to GetFileName is going to trim off all the directory information that was added either via the Path.Combine or Path.GetFullPath call.

Comment (wli3) [reply src/Tests/Microsoft.NET.Publish.Tests/PublishItemsOutputGroupOutputsTests.cs:82]: I know, but at one point the path points to a non exist file. It does not matter in the end but feels weird. I use GetFullPath to normalize the path from / to \\. GetFileName gives the wrong answer without it: GetFileName(c:\folder/file.dll) ->"folder/file.dll"

**Comment by wli3** [comment]
@marcpopMSFT preview 3 windows is passed so I should target preview 4. But preview 4 needs M2 approval. This change is "adding". It has low risk . And it would help wapproj team to try out integration few month earlier. I prepare to ask m2 approval after 1/27

**Comment by marcpopMSFT** [comment]
@wli3 preview 4 goes into M2 approval mode on Jan 27.  We just have a short M2 window for preview 3 and then VS will unlock again next week so this can go in (though noting that preview 4 is meant for low risk changes even if no approval is strictly required)

**Comment by wli3** [comment]
@marcpopMSFT so we can get approval now and get this in next week?

**Comment by marcpopMSFT** [comment]
I'm saying you don't need approval if it goes in next week.

**Comment by wli3** [comment]
@marcpopMSFT good to merge this now?

**Comment by marcpopMSFT** [comment]
@wli3 Merge when ready

**Pull request merged**